add die_with, which can run a function when dying

SeriousBug opened the pull request:
I'm pretty new to Rust so my apologies if my code isn't great. I added this for a project I'm working on. The main benefit is that with with Result types, you can include the error in your exit message. Here's how I use it in my codebase:

```rs
let conf = config::load_config(&opts)
    .die_with(|error| format!("Failed to read config file: {}", error));
```

It's also pretty easy to return an error code with this:

```rs
let conf = config::load_config(&opts)
    .die_with(|error| (format!("Failed to read config file: {}", error), 3));
```

Commit:
add "die_with" which can run a function when dying

Xavientois commented on src/lib.rs:
````diff
@@ -164,9 +164,11 @@ pub trait Die<T> {
     /// ```
     fn die_code(self, msg: &str, exit_code: i32) -> T;
+
+    fn die_with<X: PrintExit>(self, func: fn(E) -> X) -> T;
````
If you add the following test to `tests/integration_test.rs`, you'll notice that it doesn't compile:
```rust
#[test]
#[should_panic]
fn die_with_capture() {
    let code = 3;
    Result::<(), &str>::Err("error text").die_with(|error| (format!("Error is: {}", error), code));
}
```
This is because `die_with` expects a function pointer (`fn`) and therefore cannot accept closures with captures.

Instead of using a function pointer, you should generally prefer using the function traits. In this case, I would use the `FnOnce` trait. That way, you can pass closures that capture from their environment.

Xavientois commented on src/lib.rs:
```diff
@@ -192,6 +201,13 @@ impl<T> Die<T> for Option<T> {
         }
     }
+
+    fn die_with<X: PrintExit>(self, func: fn(()) -> X) -> T {
+        match self {
+            Some(t) => t,
+            None => PrintExit::print_exit(&func(()))
+        }
+    }
```
Instead of adding the `die_with` function to the `Die` trait, I would recommend just adding a new trait:
```rust
pub trait DieWith<T, E> {
    fn die_with<X: PrintExit>(self, func: fn(E) -> X) -> T;
}
```
and implementing it only for the `Result<T, E>` type. That way, you don't get the unintuitive syntax for `Option` types:
```rust
None.die_with(|()| ("error message", 2));
```

Xavientois reviewed (changes requested):
This looks awesome! Definitely a welcome addition to the library.

Before merging, it would be great to add some tests and documentation to cover your newly added function.

Thanks for contributing!

Commit:
switch to DieWith trait

Commit:
Add example to README

SeriousBug commented:
Thanks for the feedback and suggestions (and for writing this lib in the first place!)

I reverted the changes to `Die` and added `DieWith` instead, and used `FnOnce`. I also added the documentation, as well as examples for the readme. Let me know if you'd like anything else changed.

Xavientois reviewed (approved):
Documentation and changes look great!

Thanks! 😄
